include/rmm/detail/error.hpp

```diff
@@ -205,5 +207,22 @@ class out_of_range : public std::out_of_range {
                 << cudaGetErrorString(status__) << std::endl;                   \
     }                                                                           \
     assert(status__ == cudaSuccess);                                            \
-  } while (0);
+  } while (0)
+#endif
+
+/**
+ * @brief Assertion that logs a CRITICAL log message on failure.
+ */
+#ifdef NDEBUG
+#define RMM_LOGGING_ASSERT(_expr) (void)0
+#else
+#define RMM_LOGGING_ASSERT(_expr)                                                                 \
+  do {                                                                                            \
+    if (!(_expr)) {                                                                               \
+      RMM_LOG_CRITICAL(                                                                           \
+        "[" __FILE__ ":" RMM_STRINGIFY(__LINE__) "] Assertion " RMM_STRINGIFY(_expr) " failed."); \
+      rmm::logger().flush();                                                                      \
+      assert((_expr));                                                                            \
```

Comment:
This had a similar issue to that Thrust macro bug where the `_expr` was being evaluated twice. We should only evaluate it once and reuse the result. 

```suggestion
  do {                                                                                            \
    bool const success = (_expr);
    if (!success) {                                                                               \
      RMM_LOG_CRITICAL(                                                                           \
        "[" __FILE__ ":" RMM_STRINGIFY(__LINE__) "] Assertion " RMM_STRINGIFY(_expr) " failed."); \
      rmm::logger().flush();                                                                      \
      assert(success);                                                                            \
```

Reply:
Yeah I realized that the other day when the Thrust issue was found. Will fix.